cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepositoryTest.java

```diff
@@ -490,7 +508,20 @@ public void testNamespaceIsolation() throws Exception {
     }
   }
 
-  private static ClassLoader createAppClassLoader(File jarFile) throws IOException {
+  @Test
+  public void testArtifactDataset() throws IOException, ArtifactNotFoundException {
+    ArtifactDetail artifactDetail = artifactRepository.getArtifact(APP_ARTIFACT_ID);
+
+    // The artifact being created in the setupData method is using dependency tracing
+    // hence it will pick all classes under the cdap-app-fabric/src/test/main
+    Set<DatasetClass> expectedDatasets = ImmutableSet.of(
+      new DatasetClass(TrackingTable.class.getName()),
+      new DatasetClass(AppWithTxAware.PedanticTxAware.class.getName())
```

Comment:
So this tests the cases of a dataset class that extends a class which implements Dataset, and the same for a static inner class. There is also the case of a class that implements an interface that extends Dataset. That is not captured by this test, but should be tested.